xds/src/main/java/io/grpc/xds/WeightedRoundRobinLoadBalancer.java

```diff
@@ -438,110 +472,73 @@ public void shutdown() {
-        sum += child.hashCode();
+      for (SubchannelPicker picker : pickers) {
+        sum += picker.hashCode();
       }
       this.hashCode = sum
           ^ Boolean.hashCode(enableOobLoadReport)
           ^ Float.hashCode(errorUtilizationPenalty);
-
-      updateWeight();
     }
 
     @Override
     public PickResult pickSubchannel(PickSubchannelArgs args) {
-      ChildLbState childLbState = children.get(scheduler.pick());
-      WeightedChildLbState wChild = (WeightedChildLbState) childLbState;
-      PickResult pickResult = childLbState.getCurrentPicker().pickSubchannel(args);
+      int pick = scheduler.pick();
+      PickResult pickResult = pickers.get(pick).pickSubchannel(args);
       Subchannel subchannel = pickResult.getSubchannel();
       if (subchannel == null) {
         return pickResult;
       }
       if (!enableOobLoadReport) {
         return PickResult.withSubchannel(subchannel,
             OrcaPerRequestUtil.getInstance().newOrcaClientStreamTracerFactory(
-                subchannelToReportListenerMap.getOrDefault(subchannel,
-                    wChild.getOrCreateOrcaListener(errorUtilizationPenalty))));
+                reportListeners.get(pick)));
       } else {
         return PickResult.withSubchannel(subchannel);
       }
     }
 
-    private void updateWeight() {
-      float[] newWeights = new float[children.size()];
-      AtomicInteger staleEndpoints = new AtomicInteger();
-      AtomicInteger notYetUsableEndpoints = new AtomicInteger();
-      for (int i = 0; i < children.size(); i++) {
-        double newWeight = ((WeightedChildLbState) children.get(i)).getWeight(staleEndpoints,
-            notYetUsableEndpoints);
-        // TODO: add locality label once available
-        helper.getMetricRecorder()
-            .recordDoubleHistogram(ENDPOINT_WEIGHTS_HISTOGRAM, newWeight,
-                ImmutableList.of(helper.getChannelTarget()),
-                ImmutableList.of(locality));
-        newWeights[i] = newWeight > 0 ? (float) newWeight : 0.0f;
-      }
-      if (staleEndpoints.get() > 0) {
-        // TODO: add locality label once available
-        helper.getMetricRecorder()
-            .addLongCounter(ENDPOINT_WEIGHT_STALE_COUNTER, staleEndpoints.get(),
-                ImmutableList.of(helper.getChannelTarget()),
-                ImmutableList.of(locality));
-      }
-      if (notYetUsableEndpoints.get() > 0) {
-        // TODO: add locality label once available
-        helper.getMetricRecorder()
-            .addLongCounter(ENDPOINT_WEIGHT_NOT_YET_USEABLE_COUNTER, notYetUsableEndpoints.get(),
-                ImmutableList.of(helper.getChannelTarget()), ImmutableList.of(locality));
-      }
-
+    /** Returns {@code true} if weights are different than round_robin. */
+    private boolean updateWeight(float[] newWeights) {
       this.scheduler = new StaticStrideScheduler(newWeights, sequence);
-      if (this.scheduler.usesRoundRobin()) {
-        // TODO: locality label once available
```

Comment:
Should this and the other `TODO: add locality label once available` (added by Terry) be retained, but moved to the new updateWeight method?

Reply:
It was an out-of-date comment. `ImmutableList.of(locality)` is already being passed.